lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -370,112 +378,308 @@ defmodule ExDoc.Formatter.HTML.Autolink do
   `[my link Module.link/1 is here](url)`, the Module.fun/arity
   will get translated to the new href of the function.
   """
-  def elixir_functions(bin, project_funs, extension \\ ".html", lib_dirs \\ elixir_lib_dirs())
-      when is_binary(bin) do
-    bin
-    |> replace_custom_links(project_funs, extension, lib_dirs)
-    |> replace_normal_links(project_funs, extension, lib_dirs)
+  def elixir_functions(string, docs_refs, extension \\ ".html", lib_dirs \\ elixir_lib_dirs())
+      when is_binary(string) do
+    options = [
+      docs_refs: docs_refs,
+      extension: extension,
+      lib_dirs: lib_dirs
+    ]
+
+    link(string, :elixir, :function, options)
   end
 
-  module_re = Regex.source(~r{(([A-Z][A-Za-z_\d]+)\.)+})
+  @doc """
+  Helper function for autolinking elixir modules.
 
-  fun_re =
-    Regex.source(
-      ~r{([ct]:)?((#{module_re})?(([a-z_]+[A-Za-z_\d]*[\\?\\!]?)|[\{\}=&\\|\\.<>~*^@\\+\\%\\!-\/]+)/\d+)}
-    )
+  Ignores modules which are already wrapped in markdown url syntax,
+  e.g. `[Module](url)`. If the module name doesn't touch the leading
+  or trailing `]`, e.g. `[my link Module is here](url)`, the Module
+  will get translated to the new href of the module.
+  """
+  def erlang_modules(string) when is_binary(string) do
+    link(string, :erlang, :module, [])
+  end
 
-  @custom_re ~r{\[(.*?)\]\(`(#{fun_re})`\)}
-  @normal_re ~r{(?<!\[)`\s*(#{fun_re})\s*`(?!\])}
+  @doc """
+  Helper function for autolinking erlang functions.
 
-  defp replace_custom_links(bin, project_funs, extension, lib_dirs) do
-    Regex.replace(@custom_re, bin, fn all, text, match ->
-      replacement(all, match, project_funs, extension, lib_dirs, text)
-    end)
+  Only links modules that are in the Erlang distribution `lib_dir`
+  and only link functions in those modules that export a function of the
+  same name and arity.
+
+  Ignores functions which are already wrapped in markdown url syntax,
+  e.g. `[:module.test/1](url)`. If the function doesn't touch the leading
+  or trailing `]`, e.g. `[my link :module.link/1 is here](url)`, the :module.fun/arity
+  will get translated to the new href of the function.
+  """
+  def erlang_functions(string) when is_binary(string) do
+    link(string, :erlang, :function, [])
+  end
+
+  defp replace_fun(language, kind, link_type, options) do
+    case link_type do
+      :normal ->
+        fn all, match ->
+          replacement(all, language, kind, match, options)
```

Comment:
I think we should merge `replace_fun` and `replacement` such that `replace_fun` already returns a function considering the `link_type`, `language` and `kind`. :)

Reply:
Sorry, but after splitting `replacement` into two functions clauses I cannot see how to merge it with `replace_fun`

I ended up having something like this

```elixir
  defp replace_fun(language, kind, :normal, options) do
    fn all, match ->
      replacement(all, language, kind, match, options)
    end
  end

  defp replace_fun(language, kind, :custom, options) do
    fn all, text, match ->
      replacement(all, language, kind, match, text, options)
    end
  end

  # The heart of the autolinking logic
  defp replacement(string, :erlang, kind, match, text \\ nil, options) do
    lib_dirs = Keyword.get(options, :lib_dirs, default_lib_dirs(language))
    
    pmfa = {prefix, module, function, arity} = split_function(match)
    text = text || default_text(language, kind, match, pmfa)

    if doc = module_docs(:erlang, module, lib_dirs) do
      case kind do
        :module ->
          "[#{text}](#{doc}#{module}.html)"

        :function ->
          "[#{text}](#{doc}#{module}.html##{function}-#{arity})"
      end
    else
      string
    end
  end

  defp replacement(string, :elixir, kind, match, text \\ nil, options) do
    aliases = Map.get(options, :aliases, [])
    docs_refs = Map.get(options, :docs_refs, [])
    extension = Map.get(options, :extension, ".html")
    lib_dirs = Map.get(options, :lib_dirs, default_lib_dirs(language))
    locals = Map.get(options, :locals, [])
    module_id = Map.get(options, :module_id, nil)
    modules_refs = Map.get(options, :modules_refs, [])

    pmfa = {prefix, module, function, arity} = split_function(match)
    text = text || default_text(language, kind, match, pmfa)

    elixir_docs = get_elixir_docs(aliases, lib_dirs)

    case kind do
      :module ->
        cond do
          match == module_id ->
            "[`#{match}`](#{match}#{extension}#content)"

          match in modules_refs ->
            "[`#{match}`](#{match}#{extension})"

          doc = module_docs(:elixir, match, lib_dirs) ->
            "[`#{match}`](#{doc}#{match}.html)"

          true ->
            string
        end

      :function ->
        cond do
          match in locals ->
            "[#{text}](##{prefix}#{enc_h(function)}/#{arity})"

          match in docs_refs ->
            "[#{text}](#{module}#{extension}##{prefix}#{enc_h(function)}/#{arity})"

          match in @basic_type_strings ->
            "[#{text}](#{elixir_docs}#{@basic_types_page})"

          match in @built_in_type_strings ->
            "[#{text}](#{elixir_docs}#{@built_in_types_page})"

          match in @kernel_function_strings ->
            "[#{text}](#{elixir_docs}Kernel#{extension}##{prefix}#{enc_h(function)}/#{arity})"

          match in @special_form_strings ->
            "[#{text}](#{elixir_docs}Kernel.SpecialForms#{extension}##{prefix}#{
              enc_h(function)
            }/#{arity})"

          doc = module_docs(:elixir, module, lib_dirs) ->
            "[#{text}](#{doc}#{module}.html##{prefix}#{enc_h(function)}/#{arity})"

          true ->
            string
        end
      end
    end
```